### PR TITLE
86 feature forward headers

### PR DESCRIPTION
**Pull request by lukasmatusiewicz**
add headers forwarding for authproc and authsource

### CODE REVIEW

File: lib/Auth/Process/PrivacyideaAuthProc.php [hunk 1]
```diff
@@ -204,7 +212,7 @@ private function enrollToken(string $stateID, string $username): string
             $type = $this->authProcConfig['typeOfTokenToEnroll'];
             $description = "Enrolled with simpleSAMLphp";
 
-            $response = $this->pi->enrollToken($username, $genkey, $type, $description);
+            $response = $this->pi->enrollToken($username, $genkey, $type, $description, $headersToForward);
```
Comment (nilsbehlen) [inline lib/Auth/Process/PrivacyideaAuthProc.php:215]: rename $headersToForward to $headers

File: lib/Auth/Process/PrivacyideaAuthProc.php [hunk 2]
```diff
@@ -165,7 +171,7 @@ public function process(&$request): void
         // Check if it should be controlled that user has no tokens and a new token should be enrolled.
         if (!$triggered && !empty($this->authProcConfig['doEnrollToken']) && $this->authProcConfig['doEnrollToken'] === 'true')
         {
-            $stateId = $this->enrollToken($stateId, $username);
+            $stateId = $this->enrollToken($stateId, $username, $headersToForward);
```
Comment (nilsbehlen) [inline lib/Auth/Process/PrivacyideaAuthProc.php:174]: rename $headersToForward to $headers

File: lib/Auth/Process/PrivacyideaAuthProc.php [hunk 3]
```diff
@@ -147,7 +153,7 @@ public function process(&$request): void
             // Call /validate/check with a static pass from the configuration
             // This could already end up the authentication if the "passOnNoToken" policy is set.
             // Otherwise, it triggers the challenges.
-            $response = Utils::authenticatePI($state, array('otp' => $this->authProcConfig['staticPass']));
+            $response = Utils::authenticatePI($state, array('otp' => $this->authProcConfig['staticPass']), $headersToForward);
```
Comment (nilsbehlen) [inline lib/Auth/Process/PrivacyideaAuthProc.php:156]: rename $headersToForward to $headers

File: lib/Auth/Process/PrivacyideaAuthProc.php [hunk 4]
```diff
@@ -127,7 +133,7 @@ public function process(&$request): void
                     $response = null;
                     try
                     {
-                        $response = $this->pi->triggerChallenge($username);
+                        $response = $this->pi->triggerChallenge($username, $headersToForward);
```
Comment (nilsbehlen) [inline lib/Auth/Process/PrivacyideaAuthProc.php:136]: rename $headersToForward to $headers

File: lib/Auth/Source/PrivacyideaAuthSource.php [hunk 5]
```diff
@@ -219,7 +228,7 @@ public static function authSourceLogin(string $stateId, array $formParams): void
                             {
                                 try
                                 {
-                                    $response = $source->pi->triggerChallenge($username);
+                                    $response = $source->pi->triggerChallenge($username, $headersToForward);
```
Comment (nilsbehlen) [inline lib/Auth/Source/PrivacyideaAuthSource.php:231]: rename $headersToForward to $headers

File: lib/Auth/Source/PrivacyideaAuthSource.php [hunk 6]
```diff
@@ -237,7 +246,7 @@ public static function authSourceLogin(string $stateId, array $formParams): void
 
                             try
                             {
-                                $response = $source->pi->validateCheck($username, $password);
+                                $response = $source->pi->validateCheck($username, $password, $headersToForward);
```
Comment (nilsbehlen) [inline lib/Auth/Source/PrivacyideaAuthSource.php:249]: rename $headersToForward to $headers

File: lib/Auth/Source/PrivacyideaAuthSource.php [hunk 7]
```diff
@@ -249,7 +258,7 @@ public static function authSourceLogin(string $stateId, array $formParams): void
                             Logger::error("privacyIDEA: Invalid authentication flow. Please set 'authenticationFlow' to one of the following values: 'sendPass', 'triggerChallenge' or 'separateOTP'. Fallback to default (sendPass)");
                             try
                             {
-                                $response = $source->pi->validateCheck($username, $password);
+                                $response = $source->pi->validateCheck($username, $password, $headersToForward);
```
Comment (nilsbehlen) [inline lib/Auth/Source/PrivacyideaAuthSource.php:261]: rename $headersToForward to $headers

File: lib/Auth/Source/PrivacyideaAuthSource.php [hunk 8]
```diff
@@ -266,7 +275,7 @@ public static function authSourceLogin(string $stateId, array $formParams): void
         {
             try
             {
-                $response = Utils::authenticatePI($state, $formParams);
+                $response = Utils::authenticatePI($state, $formParams, $headersToForward);
```
Comment (nilsbehlen) [inline lib/Auth/Source/PrivacyideaAuthSource.php:278]: rename $headersToForward to $headers

File: lib/Auth/Utils.php [hunk 9]
```diff
@@ -21,7 +21,7 @@ class Utils
      * @return PIResponse|null An array containing attributes and detail, or NULL.
      * @throws Exception
      */
-    public static function authenticatePI(array &$state, array $formParams): ?PIResponse
+    public static function authenticatePI(array &$state, array $formParams, array $headersToForward): ?PIResponse
```
Comment (nilsbehlen) [inline lib/Auth/Utils.php:24]: rename $headersToForward to $headers

File: lib/Auth/Utils.php [hunk 10]
```diff
@@ -76,7 +76,7 @@ public static function authenticatePI(array &$state, array $formParams): ?PIResp
                     // If the authentication has been confirmed on the phone, the authentication has to be finalized with a
                     // call to /validate/check with an empty pass
                     // https://privacyidea.readthedocs.io/en/latest/tokens/authentication_modes.html#outofband-mode
-                    $response = $pi->validateCheck($username, "", $transactionID);
+                    $response = $pi->validateCheck($username, "", $transactionID, $headersToForward);
```
Comment (nilsbehlen) [inline lib/Auth/Utils.php:79]: rename $headersToForward to $headers

File: lib/Auth/Utils.php [hunk 11]
```diff
@@ -96,7 +96,7 @@ public static function authenticatePI(array &$state, array $formParams): ?PIResp
             {
                 try
                 {
-                    $response = $pi->validateCheckU2F($username, $transactionID, $u2fSignResponse);
+                    $response = $pi->validateCheckU2F($username, $transactionID, $u2fSignResponse, $headersToForward);
```
Comment (nilsbehlen) [inline lib/Auth/Utils.php:99]: rename $headersToForward to $headers

File: lib/Auth/Utils.php [hunk 12]
```diff
@@ -117,7 +117,7 @@ public static function authenticatePI(array &$state, array $formParams): ?PIResp
             {
                 try
                 {
-                    $response = $pi->validateCheckWebAuthn($username, $transactionID, $webAuthnSignResponse, $origin);
+                    $response = $pi->validateCheckWebAuthn($username, $transactionID, $webAuthnSignResponse, $origin, $headersToForward);
```
Comment (nilsbehlen) [inline lib/Auth/Utils.php:120]: rename $headersToForward to $headers

File: lib/Auth/Utils.php [hunk 13]
```diff
@@ -129,7 +129,7 @@ public static function authenticatePI(array &$state, array $formParams): ?PIResp
         {
             try
             {
-                $response = $pi->validateCheck($username, $formParams["otp"], $transactionID);
+                $response = $pi->validateCheck($username, $formParams["otp"], $transactionID, $headersToForward);
```
Comment (nilsbehlen) [inline lib/Auth/Utils.php:132]: rename $headersToForward to $headers

File: lib/Auth/Utils.php [hunk 14]
```diff
@@ -424,4 +424,37 @@ public static function getClientIP(): ?string
         Logger::debug('privacyIDEA: client ip: ' . $result);
         return $result;
     }
+
+    /**
+     * Search for the wished headers and return all found with their values.
```
[the remark this answers is not in the document]
Comment (lukasmatusiewicz) [reply lib/Auth/Utils.php]: fixed [252d981](https://github.com/privacyidea/simplesamlphp-module-privacyidea/pull/198/commits/252d98129a749321bb0a53eb533385240a941df4)

File: lib/Auth/Utils.php (hunk 13 above)
Comment (lukasmatusiewicz) [reply lib/Auth/Utils.php]: fixed [f2bb187](https://github.com/privacyidea/simplesamlphp-module-privacyidea/pull/198/commits/f2bb187f584e1a7a8c5c5fec38dee8d19a3e6b78)

File: lib/Auth/Utils.php (hunk 12 above)
Comment (lukasmatusiewicz) [reply lib/Auth/Utils.php]: fixed [f2bb187](https://github.com/privacyidea/simplesamlphp-module-privacyidea/pull/198/commits/f2bb187f584e1a7a8c5c5fec38dee8d19a3e6b78)

File: lib/Auth/Utils.php (hunk 11 above)
Comment (lukasmatusiewicz) [reply lib/Auth/Utils.php]: fixed [f2bb187](https://github.com/privacyidea/simplesamlphp-module-privacyidea/pull/198/commits/f2bb187f584e1a7a8c5c5fec38dee8d19a3e6b78)

File: lib/Auth/Utils.php (hunk 10 above)
Comment (lukasmatusiewicz) [reply lib/Auth/Utils.php]: fixed [f2bb187](https://github.com/privacyidea/simplesamlphp-module-privacyidea/pull/198/commits/f2bb187f584e1a7a8c5c5fec38dee8d19a3e6b78)

File: lib/Auth/Utils.php (hunk 9 above)
Comment (lukasmatusiewicz) [reply lib/Auth/Utils.php]: fixed [f2bb187](https://github.com/privacyidea/simplesamlphp-module-privacyidea/pull/198/commits/f2bb187f584e1a7a8c5c5fec38dee8d19a3e6b78)

File: lib/Auth/Process/PrivacyideaAuthProc.php (hunk 2 above)
Comment (lukasmatusiewicz) [reply lib/Auth/Process/PrivacyideaAuthProc.php]: fixed [f2bb187](https://github.com/privacyidea/simplesamlphp-module-privacyidea/pull/198/commits/f2bb187f584e1a7a8c5c5fec38dee8d19a3e6b78)

File: lib/Auth/Process/PrivacyideaAuthProc.php [hunk 15]
```diff
@@ -186,10 +192,12 @@ public function process(&$request): void
...
      * @param string $stateID
      * @param string $username
+     * @param array $headersToForward
      * @return string
-     * @throws PIBadRequestException|NoState
+     * @throws NoState
+     * @throws PIBadRequestException
      */
-    private function enrollToken(string $stateID, string $username): string
+    private function enrollToken(string $stateID, string $username, $headersToForward): string
```
[the remark this answers is not in the document]
Comment (lukasmatusiewicz) [reply lib/Auth/Process/PrivacyideaAuthProc.php:200]: fixed [f2bb187](https://github.com/privacyidea/simplesamlphp-module-privacyidea/pull/198/commits/f2bb187f584e1a7a8c5c5fec38dee8d19a3e6b78)

File: lib/Auth/Process/PrivacyideaAuthProc.php (hunk 1 above)
Comment (lukasmatusiewicz) [reply lib/Auth/Process/PrivacyideaAuthProc.php:215]: fixed [f2bb187](https://github.com/privacyidea/simplesamlphp-module-privacyidea/pull/198/commits/f2bb187f584e1a7a8c5c5fec38dee8d19a3e6b78)

File: lib/Auth/Source/PrivacyideaAuthSource.php [hunk 16]
```diff
@@ -203,7 +212,7 @@ public static function authSourceLogin(string $stateId, array $formParams): void
                         Logger::error("privacyIDEA: Authentication flow not found in the config file. Please add the 'authenticationFlow' with one of the following values: 'sendPass', 'triggerChallenge' or 'separateOTP'. Until then, the login mask contains per default 1 user field and 1 pass field.");
                         try
                         {
-                            $response = $source->pi->validateCheck($username, $password);
+                            $response = $source->pi->validateCheck($username, $password, $headersToForward);
```
[the remark this answers is not in the document]
Comment (lukasmatusiewicz) [reply lib/Auth/Source/PrivacyideaAuthSource.php]: fixed [f2bb187](https://github.com/privacyidea/simplesamlphp-module-privacyidea/pull/198/commits/f2bb187f584e1a7a8c5c5fec38dee8d19a3e6b78)

**Review by nilsbehlen — APPROVED** [review]

**Pull request merged**